drivers/bus/acpi/compbatt/compbatt.c

```diff
@@ -202,40 +526,665 @@ CompBattDisableStatusNotify(IN PCOMPBATT_DEVICE_EXTENSION DeviceExtension)
 
     /* Done */
     ExReleaseFastMutex(&DeviceExtension->Lock);
-    if (CompBattDebug & 0x100) DbgPrint("CompBatt: EXITING DisableStatusNotify\n");
+    if (CompBattDebug & COMPBATT_DEBUG_FUNC_ENTER_EXIT) DbgPrint("CompBatt: EXITING DisableStatusNotify\n");
     return STATUS_SUCCESS;
 }
 
+/**
+ * @brief
+ * Calculates the total discharging/charging rate flow of each individual
+ * battery linked with the composite battery and determines whether at
+ * least one battery is behaving improperly.
+ *
+ * @param[in] DeviceExtension
+ * A pointer to a device extension which describes the composite battery
+ * itself. It is used to gather each connected battery in the list with
+ * the composite battery.
+ *
+ * @param[out] TotalRate
+ * A pointer returned to caller that describes the total accumulated
+ * rate flow of all batteries.
+ *
+ * @param[out] BatteriesCount
+ * A pointer returned to caller that describes the batteries present.
+ *
+ * @return
+ * Returns TRUE if at least one battery is behaving improperly, FALSE
+ * otherwise. This is determined by the fact if a battery has a negative
+ * rate but is charging, or if it has a positive rate but is discharging.
+ */
+static
+BOOLEAN
+CompBattCalculateTotalRateAndLinkedBatteries(
+    _In_ PCOMPBATT_DEVICE_EXTENSION DeviceExtension,
+    _Out_ PULONG TotalRate,
+    _Out_ PULONG BatteriesCount)
+{
+    PCOMPBATT_BATTERY_DATA BatteryData;
+    PLIST_ENTRY ListHead, NextEntry;
+    BOOLEAN BadBattery = FALSE;
+    ULONG LinkedBatteries = 0;
+    ULONG BadBatteriesCount = 0;
+    ULONG ComputedRate = 0;
+
+    /* Loop over the linked batteries and sum up the total capacity rate */
+    ExAcquireFastMutex(&DeviceExtension->Lock);
+    ListHead = &DeviceExtension->BatteryList;
+    NextEntry = ListHead->Flink;
+    while (NextEntry != ListHead)
+    {
+        /* Acquire the remove lock so this battery does not disappear under us */
+        BatteryData = CONTAINING_RECORD(NextEntry, COMPBATT_BATTERY_DATA, BatteryLink);
+        if (NT_SUCCESS(IoAcquireRemoveLock(&BatteryData->RemoveLock, BatteryData->Irp)))
```

Comment:
You could then do:
```suggestion
        if (!NT_SUCCESS(IoAcquireRemoveLock(&BatteryData->RemoveLock, BatteryData->Irp)))
            continue;
```
and reduce the indent level of the whole code below.